Tests/NativeBuilderTests/ContainerBuildSnapshotterTests/DiffKeyEdgeCasesTests.swift

```diff
@@ -0,0 +1,637 @@
+    }
+
+    private func createUnixSocket(at url: URL) throws {
+        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
+
+        let fd = socket(AF_UNIX, SOCK_STREAM, 0)
+        if fd < 0 { throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno), userInfo: [NSLocalizedDescriptionKey: "socket(AF_UNIX) failed"]) }
+        defer { _ = close(fd) }
+
+        var addr = sockaddr_un()
+        addr.sun_family = sa_family_t(AF_UNIX)
+
+        // Copy UTF-8 C string into sun_path
+        let pathBytes = url.path.utf8CString
+        let maxLen = MemoryLayout.size(ofValue: addr.sun_path)
+        #if os(Linux)
+        // On Linux the struct often allows ~108 chars including NUL.
+        #else
+        // On Darwin it's typically 104 including NUL.
+        #endif
+        if pathBytes.count > maxLen {
+            return
+        }
+
+        withUnsafeMutablePointer(to: &addr.sun_path) { ptr in
+            let raw = UnsafeMutableRawPointer(ptr).assumingMemoryBound(to: CChar.self)
+            _ = pathBytes.withUnsafeBytes { src in
+                memcpy(raw, src.baseAddress, pathBytes.count)
+            }
+        }
+
+        let len = socklen_t(MemoryLayout<sockaddr_un>.size)
+        let bindRes: Int32 = withUnsafePointer(to: &addr) {
+            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) { p in
+                bind(fd, p, len)
+            }
+        }
+        if bindRes != 0 {
+            throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno), userInfo: [NSLocalizedDescriptionKey: "bind(AF_UNIX) failed"])
+        }
+    }
+
+    private func createCharDevice(_ url: URL, major: UInt32, minor: UInt32, perms: Int = 0o644) throws {
+        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
+        #if os(Linux)
+        let dev: dev_t = makedev(major, minor)
+        let mode: mode_t = S_IFCHR | mode_t(perms)
+        let r: Int32 = url.path.withCString { mknod($0, mode, dev) }
+        #else
+        // Swift cannot import function-like macro makedev on Darwin. Compose dev_t manually:
+        // On Darwin: makedev(x, y) = ((x << 24) | (y)).
+        let devRaw: UInt32 = (major << 24) | (minor & 0x00FF_FFFF)
+        #if arch(x86_64) || arch(arm64)
+        let dev: dev_t = dev_t(Int32(bitPattern: devRaw))
+        #else
+        let dev: dev_t = dev_t(devRaw)
+        #endif
+        let mode: mode_t = S_IFCHR | mode_t(perms)
+        let r: Int32 = url.path.withCString { mknod($0, mode, dev) }
+        #endif
+        if r != 0 { throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno), userInfo: [NSLocalizedDescriptionKey: "mknod failed"]) }
+    }
+
+    private func computeKey(base: URL?, target: URL, options: InspectorOptions? = nil) async throws -> DiffKey {
```

Comment:
We have `computeKey` and `computeDiffKey` that do very similar things. Is there a way we can take out some of the utility functions and put them in a common place between these tests?

Reply:
yeah, this a good idea!

Reply:
Hey @wlan0 were you planning to make this change in this PR?

Reply:
Yeah I missed this one in my last push